plugin-server/src/main/ingestion-queues/session-recordings-consumer.ts

```diff
@@ -91,78 +92,70 @@ export const startSessionRecordingEventsConsumer = async ({
 
-            // Get the offsets for the last message for each partition, from
-            // messages
-            const offsets = messages.reduce((acc, message) => {
-                if (!message.partition || !message.offset) {
-                    return acc
-                }
-                const partition = message.partition.toString()
-                const offset = message.offset.toString()
-                if (!acc[partition] || acc[partition] < offset) {
-                    acc[partition] = offset
-                }
-                return acc
-            }, {} as Record<string, string>)
-
-            const topicPartitionOffsets = Object.entries(offsets).map(([partition, offset]) => ({
-                topic: KAFKA_SESSION_RECORDING_EVENTS,
-                partition: parseInt(partition, 10),
-                offset: parseInt(offset, 10),
-            }))
+            // We try to process the entire batch at once. Connections to
+            // PostgreSQL should be pooled, so there shouldn't be too much
+            // danger of overloading the database. It's possible that we end up
+            // hitting the Producers internal queue limit.
+            // TODO: handle either limiting concurrency, or handle the errors
+            // that would result.
+            await Promise.all(messages.map((message: Message) => eachMessageWithContext(message)))
 
-            consumer.offsetsStore(topicPartitionOffsets)
+            // On each loop, we flush the producer to ensure that all messages
+            // are sent to Kafka.
+            await flushProducer(producer)
+            storeOffsetsForMessages(messages, consumer)
         }
     }
 
-    void startConsuming()
+    const mainLoop = startConsuming()
 
     const isHealthy = async () => {
-        // TODO: add health check implementation
+        // If the consumer is in a health state, return true. Otherwise, return
+        // false.
+        consumer
         return true
     }
 
     const stop = async () => {
         status.info('🔁', 'Stopping session recordings consumer')
-        await new Promise((resolve, reject) => {
-            consumer.disconnect((error, data) => {
-                if (error) {
-                    status.error('🔥', 'Failed to disconnect session recordings consumer', { error })
-                    reject(error)
-                } else {
-                    status.info('🔁', 'Disconnected session recordings consumer')
-                    resolve(data)
-                }
-            })
-        })
 
-        await flushProducer(producer)
+        // First we signal to the mainLoop that we should be stopping. The main
+        // loop should complete one loop, flush the producer, and store it's offsets.
+        isRunning = false
+
+        // Wait for the main loop to finish, but only give it 10 seconds
+        await Promise.race([mainLoop, new Promise((resolve) => setTimeout(resolve, 10000))])
```

Comment:
tangent, no action needed: our current `terminationGracePeriodSeconds` is 30 seconds, but we can increase it to a minute if needed, to increase this timeout.
Does not seem needed for recordings though, as the time to process one batch should be two orders of magnitude lower indeed.